Enforce Strict `@types` Dependencies

ObliviousHarmony opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

As part of my work on #37280, I will be centralizing our TypeScript configuration to make maintenance easier. In the process of doing that, however, I found that there were some `@types` declarations missing from a few packages. By default, TypeScript will search for types in `./node_modules` and will traverse parent directories looking for the same. This can cause issues when a `@types` declaration is missing from the package but a parent directory has one from a different version of the package. 

This pull request isolates the types to the project's directory, and as a consequence, caused a number of type errors across the monorepo. This pull request contains fixes for all of those errors.

In the interesting of an easier time reviewing and testing, this pull request does not contain any TypeScript refactoring changes. The changes in this PR are **just** the result of fixing any errors introduced by setting `"typeRoots"` in all of the `tsconfig.json` and `tsconfig-cjs.json` files.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

* Verify that all of the CI checks pass.
* Verify that the output of `pnpm run test --filter="!woocommerce"` matches what was present before this PR.
* Verify that WooCommerce's React admin pages still work as expected.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

ObliviousHarmony committed:
Fixed Turborepo Deprecation

The use of environment variables in `"dependsOn"` has been deprecated.
This was replaced by an `"env"` property.

ObliviousHarmony committed:
Added Missing `@types` Packages

As part of refactoring the TypeScript configuration, all
of the `"typeRoots"` will be set to the local package folder.
The impact here is that any types found on accident will
no longer be picked up.

ObliviousHarmony committed:
Localized TSConfig Type Roots

The default behavior is for TypeScript to search through all of
the accessible `node_modules` folders in parent directories. This
commit changes the `"typeRoots"` configuration to only look at
local directories.

ObliviousHarmony committed:
Fixed Miscellaneous TypeScript Errors

After localizing the `"typeRoots"` there were some problems introduced.
This commit addresses those problems so that the build compiles.

ObliviousHarmony committed:
Added Missing `@types/node`

ObliviousHarmony committed:
Centralized `@woocommerce/internal-js-tests` Package

As part of the package isolation work here this commit loads the
Jest preset from `node_modules` instead of a monorepo reference.

ObliviousHarmony committed:
Added Changelogs

ObliviousHarmony committed:
Updated Lock File

ObliviousHarmony committed:
Fixed Linting Errors

ilyasfoo commented on packages/js/components/src/utils.tsx:
```diff
@@ -64,7 +64,7 @@ function createOrderedChildren< T = Fill.Props, S = Record< string, unknown > >(
 	props: T,
 	injectProps?: S
-) {
+): React.ReactSVGElement {
```
I think this is unintended? `React.ReactElement` seems to be the better suited type
```suggestion
): React.ReactElement {
```

ObliviousHarmony replied:
I went with what VS Code told me that the function would be returning 😄 Thanks for the tip here, I'll make the change!

ilyasfoo commented on packages/js/product-editor/src/utils/create-ordered-children.tsx:
```diff
@@ -64,7 +64,7 @@ function createOrderedChildren< T = Fill.Props, S = Record< string, unknown > >(
 	props: T,
 	injectProps?: S
-) {
+): React.ReactSVGElement {
```
Same as above

```suggestion
): React.ReactElement {
```

chihsuan reviewed (dismissed):
Awesome! Looks good to me and tested well. 👍

ilyasfoo reviewed (changes requested):
Thanks for working on this, @ObliviousHarmony! This tested well, I have a minor comment:

louwie17 commented on packages/js/tsconfig-cjs.json:
```diff
@@ -1,7 +1,7 @@
 	"extends": "./tsconfig",
 	"compilerOptions": {
-		"target": "es6",
+		"target": "esnext",
```
This is quite the jump, not that I am opposed. But will this cause any issues for the newest release of packages, for the plugins that bundle it within their code?
We did probably have to create major release versions for each new package or not?

louwie17 replied:
Although it may be fine anyhow given the `module` hasn't changed - my `tsconfig` knowledge is quite limited.

ObliviousHarmony replied:
Ultimately the goal of the `target` option is to tell TypeScript what features the output JavaScript should be able to make use of. On the browser front, Gutenberg's `tsconfig` files also target `"esnext"` and their Babel configuration targets browsers that WordPress has declared compatibility for.

With that said, however, I think you're right that this is a bit heavy-handed. I'll change this to `"es2019"` which is the same as our `tsconfig.json`. This is supported by the version of Node we've declared compatibility with, as well as all of the browsers that we are compatible with. Given that our ESM builds already target this I don't think it requires a major version release.

louwie17 reviewed (commented):
Looks good from my end, aside from the two items that @ilyasfoo has already pointed out around the `ReactElement` type.

ObliviousHarmony committed:
Aligned `tsconfig-cjs.json` `"target"`

Using `"esnext"` was a little heavy-handed so I've changed the target
to match the ESM build's target. Given that ES2019 is supported by
the Node version we've declared compatibility with, I don't think this
will be a problem.

ObliviousHarmony committed:
Made Type Less Explicit

I went with what VS Code told me the return type was, but, apparently
there's a better type here!

ObliviousHarmony committed:
Updated Lock FIle

ilyasfoo reviewed (dismissed):
LGTM! 🚢

ObliviousHarmony committed:
Merge branch 'trunk' into fix/typescript-package-isolation

ObliviousHarmony committed:
Updated Lock File

ObliviousHarmony committed:
Fixed Type Error

ObliviousHarmony committed:
Fixed `moment` Type Error

ilyasfoo reviewed (approved):
Changes lgtm and tests well! 🚢
